src/runtime/cuda/cuda_device_api.cc

```diff
@@ -214,10 +234,33 @@ class CUDADeviceAPI final : public DeviceAPI {
   bool PinData(void* ptr, size_t nbytes) override {
     // prevent users from pinning empty tensors or graphs
     if (ptr == nullptr || nbytes == 0) return false;
+    TensorDispatcher* tensor_dispatcher = TensorDispatcher::Global();
+    // Minimize the pinned memory pool allocated by backend (via tensoradapter)
+    // to preserve enough memory for DGL inherited in-place pin-memory operation
+    if (tensor_dispatcher->IsAvailable()) {
+      tensor_dispatcher->CUDAHostAllocEmptyCache();
```

Comment:
The function is named `AllocEmptyCache` but the comment says "Minimize the pinned memory pool".  At a first glance I don't understand the connection.  I need to look into the tensor adapter code to figure out this is actually "empty the cache in host allocator" rather than "allocating an empty cache".  So it's better to change the name to `CUDAHostAllocatorEmptyCache`.

Also, does it mean we are emptying the cache every time a graph object is pinned?  Would it incur additional overhead?

Reply:
>So it's better to change the name to CUDAHostAllocatorEmptyCache.

Thanks. I updated accordingly.

> Also, does it mean we are emptying the cache every time a graph object is pinned? Would it incur additional overhead?

Yes, we need to empty the cache using cudaHostFree() when a graph is pinned by DGL, so it would add extra overhead. I think it is fine.  In DGL, pinning/unpinning a graph via `cudaHostRegister`/`cudaHostUnRegister` is very expensive, so we normally do it once for the entire graph structure and feature storage at the beginning. 
Compared to the cost of pinning the entire graph, the overhead of empty cache should be small. And `cudaHostRegister` should not be invoked often, otherwise, we should cache it via `CUDARawHostAlloc`.  It is also necessary to do so, as we need to isolate two pinned pools and reserve enough space for `cudaHostRegister`, in preparation for the entire graph/feature.

Reply:
>If this is to be used for pinning sampled subgraphs, would it kind of cancel the benefit of asynchronous transfer of subgraph structures?

When pinning sampled subgraphs, we should never call this function `PinData`, which uses `cudaHostRegister`. Instead, for small subgraphs (pinning/destroying frequently), we should use `CUDARawHostAlloc` imported by this PR.

To sum up, 
(1) in UVA mode, we pin the entire graph and feature ONCE with DGL's `PinData()` function via `dgl.graph.pin_memory_()`. Before pinning, we clean the backend's pinned mem pool by `CUDAHostAllocatorEmptyCache`.
(2) in non-UVA mode, we can pin the subgraph for each batch with a different function `AllocPinnedDataSpace()` via `dgl.graph._graph.pin_memory()` and it shares the same pinned memory pool managed by PyTorch. It does not need to empty cache 'manually' by DGL, but handled by PyTorch.

Reply:
I see.  Because `PinData` is pinning the data in-place with `CUDAHostRegister`, we have to empty the cache before pinning, and `CUDAHostRegister` is not called often anyway.

I think we should rename this `PinData` function as it is now ambiguous: the name could refer to both in-place pinning and out-of-place pinning.  Something more explicit like `PinDataInPlace` or `RegisterHostMemory` should be good.